cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,378 @@
+                self._custom_gate_families.append(g)
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def with_params(
+        self,
+        *,
+        name: Optional[str] = None,
+        unroll_circuit_op: Optional[bool] = None,
+        accept_global_phase: Optional[bool] = None,
+    ) -> 'Gateset':
+        """Returns a copy of this Gateset with identical gates and new values for named arguments.
+
+        If a named argument is None then corresponding value of this Gateset is used instead.
+
+        Args:
+            name: New name for the Gateset.
+            unroll_circuit_op: If True, new Gateset will recursively validate
+                `cirq.CircuitOperation` by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, new Gateset will accept `cirq.GlobalPhaseOperation`.
+
+        Returns:
+            `self` if all new values are None or identical to the values of current Gateset.
+            else a new Gateset with identical gates and new values for named arguments.
+        """
+
+        def val_if_none(var: Any, val: Any) -> Any:
+            return var if var is not None else val
+
+        name = val_if_none(name, self._name)
+        unroll_circuit_op = val_if_none(unroll_circuit_op, self._unroll_circuit_op)
+        accept_global_phase = val_if_none(accept_global_phase, self._accept_global_phase)
+        if (
+            name == self._name
+            and unroll_circuit_op == self._unroll_circuit_op
+            and accept_global_phase == self._accept_global_phase
+        ):
+            return self
+        return Gateset(
+            *self.gates,
+            name=name,
+            unroll_circuit_op=cast(bool, unroll_circuit_op),
+            accept_global_phase=cast(bool, accept_global_phase),
+        )
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of a given Gate/Operation in this Gateset.
+
+        Containment checks are handled as follows:
+            a) For Gates or Operations that have an underlying gate (i.e. op.gate is not None):
+                - Forwards the containment check to the underlying GateFamily's
+                - Examples of such operations include `GateOperations` and their controlled and
+                    tagged variants (i.e. instances of `TaggedOperation`, `ControlledOperation`
+                    where `op.gate` is not None) etc.
+
+            b) For Operations that do not have an underlying gate:
+                - Forwards the containment check to `self._validate_operation(item)`.
+                - Examples of such operations include `CircuitOperations` and their controlled
+                    and tagged variants (i.e. instances of `TaggedOperation`, `ControlledOperation`
+                    where `op.gate` is None) etc.
```

Comment:
Missing a few links in here where `TaggedOperation` -> `cirq.TaggedOperation` for docs linking.

Reply:
done